RecoLocalTracker/SiPixelRecHits/plugins/PixelRecHits.cu

```diff
@@ -0,0 +1,86 @@
+// C++ headers
+#include <algorithm>
+#include <numeric>
+
+// CUDA runtime
+#include <cuda_runtime.h>
+
+// CMSSW headers
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/cudaCheck.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/device_unique_ptr.h"
+#include "RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterGPUKernel.h"
+#include "RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusteringConstants.h"
+
+#include "PixelRecHits.h"
+#include "gpuPixelRecHits.h"
+
+namespace {
+  __global__ void setHitsLayerStart(uint32_t const* __restrict__ hitsModuleStart,
+                                    pixelCPEforGPU::ParamsOnGPU const* cpeParams,
+                                    uint32_t* hitsLayerStart) {
+    auto i = blockIdx.x * blockDim.x + threadIdx.x;
+
+    assert(0 == hitsModuleStart[0]);
+
+    if (i < 11) {
+      hitsLayerStart[i] = hitsModuleStart[cpeParams->layerGeometry().layerStart[i]];
+#ifdef GPU_DEBUG
+      printf("LayerStart %d %d: %d\n", i, cpeParams->layerGeometry().layerStart[i], hitsLayerStart[i]);
+#endif
+    }
+  }
+}  // namespace
+
+namespace pixelgpudetails {
+
+  TrackingRecHit2DCUDA PixelRecHitGPUKernel::makeHitsAsync(SiPixelDigisCUDA const& digis_d,
+                                                           SiPixelClustersCUDA const& clusters_d,
+                                                           BeamSpotCUDA const& bs_d,
+                                                           pixelCPEforGPU::ParamsOnGPU const* cpeParams,
+                                                           cudaStream_t stream) const {
+    auto nHits = clusters_d.nClusters();
+    TrackingRecHit2DCUDA hits_d(nHits, cpeParams, clusters_d.clusModuleStart(), stream);
+
+    int threadsPerBlock = 128;
+    int blocks = digis_d.nModules();  // active modules (with digis)
+
+#ifdef GPU_DEBUG
+    std::cout << "launching getHits kernel for " << blocks << " blocks" << std::endl;
+#endif
+    if (blocks)  // protect from empty events
+      gpuPixelRecHits::getHits<<<blocks, threadsPerBlock, 0, stream>>>(
+          cpeParams, bs_d.data(), digis_d.view(), digis_d.nDigis(), clusters_d.view(), hits_d.view());
+    cudaCheck(cudaGetLastError());
+#ifdef GPU_DEBUG
+    cudaDeviceSynchronize();
+    cudaCheck(cudaGetLastError());
+#endif
+
+    // assuming full warp of threads is better than a smaller number...
+    if (nHits) {
+      setHitsLayerStart<<<1, 32, 0, stream>>>(clusters_d.clusModuleStart(), cpeParams, hits_d.hitsLayerStart());
+      cudaCheck(cudaGetLastError());
+    }
+
+    if (nHits >= TrackingRecHit2DSOAView::maxHits()) {
+      edm::LogWarning("PixelRecHitGPUKernel")
```

Comment:
Looks like the MessageLogger is now using c++17 features (namely, `std::string_view`) so we cannot use in a .cu file.

@makortel @VinInn what do you think ?
Should we keep only the kernels in a .cu file, and move the host code to a .cc file ?

Reply:
I would say remove the warning

Reply:
I've coded a possible workaround in https://github.com/cms-patatrack/cmssw/pull/421 .